src/types/excelsheet.rs

```diff
@@ -76,6 +79,100 @@ impl Pagination {
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) enum SelectedColumns {
+    All,
+    ByIndex(Vec<usize>),
+    ByName(Vec<String>),
+}
+
+impl SelectedColumns {
+    pub(crate) fn validate_columns(&self, column_names: &[String]) -> FastExcelResult<()> {
+        match self {
+            SelectedColumns::All => Ok(()),
+            // If no selected indice is >= to the len of column_names, we're good
+            SelectedColumns::ByIndex(indices) => indices.iter().try_for_each(|idx| {
+                if idx >= &column_names.len() {
+                    Err(FastExcelErrorKind::ColumnNotFound(IdxOrName::Idx(*idx)).into())
+                } else {
+                    Ok(())
+                }
+            }),
+            // Every selected column must be in the provided column_names
+            SelectedColumns::ByName(selected_names) => {
+                selected_names.iter().try_for_each(|selected_name| {
+                    if column_names.contains(selected_name) {
+                        Ok(())
+                    } else {
+                        Err(FastExcelErrorKind::ColumnNotFound(IdxOrName::Name(
+                            selected_name.to_string(),
+                        ))
+                        .into())
+                    }
+                })
+            }
+        }
+    }
+
+    pub(crate) fn idx_for_column(
+        &self,
+        col_names: &[String],
+        col_name: &str,
+        col_idx: usize,
+    ) -> Option<usize> {
+        match self {
+            SelectedColumns::All => None,
+            SelectedColumns::ByIndex(indices) => {
+                if indices.contains(&col_idx) {
+                    Some(col_idx)
+                } else {
+                    None
+                }
+            }
+            SelectedColumns::ByName(names) => {
+                // cannot use .contains() because we have &String and &str
+                if names.iter().any(|name| name == col_name) {
+                    col_names.iter().position(|name| name == col_name)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+impl TryFrom<Option<&PyList>> for SelectedColumns {
+    type Error = FastExcelError;
+
+    fn try_from(value: Option<&PyList>) -> FastExcelResult<Self> {
+        use FastExcelErrorKind::InvalidParameters;
+
+        match value {
+            None => Ok(Self::All),
+            Some(py_list) => {
+                if let Ok(name_vec) = py_list.extract::<Vec<String>>() {
+                    if name_vec.is_empty() {
+                        Err(InvalidParameters("list of select columns is empty".to_string()).into())
+                    } else {
+                        Ok(Self::ByName(name_vec))
+                    }
+                } else if let Ok(index_vec) = py_list.extract::<Vec<usize>>() {
+                    if index_vec.is_empty() {
+                        Err(InvalidParameters("list of select columns is empty".to_string()).into())
```

Comment:
In theory I guess we'll never be in that case with the first check on empty string list

Reply:
yeah, the check is cumbersome anyway... I'll just use https://docs.rs/pyo3/latest/pyo3/types/struct.PyList.html#method.len once instead

Reply:
bbb3542

Reply:
I prefer this approach thanks 👍